internal/redfishwrapper/firmware.go

```diff
@@ -0,0 +1,439 @@
+package redfishwrapper
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/bmc-toolbox/bmclib/v2/constants"
+	bmclibErrs "github.com/bmc-toolbox/bmclib/v2/errors"
+)
+
+type installMethod string
+
+const (
+	unstructuredHttpPush installMethod = "unstructuredHttpPush"
+	multipartHttpUpload  installMethod = "multipartUpload"
+)
+
+var (
+	errMultiPartPayload   = errors.New("error preparing multipart payload")
+	errUpdateParams       = errors.New("error in redfish UpdateParameters payload")
+	errTaskIdFromRespBody = errors.New("failed to identify firmware install taskID from response body")
+)
+
+type RedfishUpdateServiceParameters struct {
+	Targets            []string                     `json:"Targets"`
+	OperationApplyTime constants.OperationApplyTime `json:"@Redfish.OperationApplyTime"`
+	Oem                json.RawMessage              `json:"Oem"`
+}
+
+// FirmwareUpload uploads and initiates the firmware install process
+func (c *Client) FirmwareUpload(ctx context.Context, reader io.Reader, params *RedfishUpdateServiceParameters) (taskID string, err error) {
+	// limit to *os.File until theres a need for other types of readers
+	updateFile, ok := reader.(*os.File)
```

Comment:
This feels like a mistake.

Reply:
Based on our discussion, I agree and the changes are in the child branch https://github.com/bmc-toolbox/bmclib/pull/363/commits/3a43b9230c14eb9d6fd634f544bfa58b254d47a9